### ISSUE TITLE
Parsing and display problems in MFHD holdings data

### ISSUE THREAD

**Issue by gibsonjc** [CARLI/vufind#247]
Sometimes I see other MFHD fields/subfields displaying in the "Volume Holdings" area of the Holdings tab like the system is incorrectly parsing the data.

Not sure if it is related in part to whatever mechanism is artificially combining MFHD data for display #246 ?  (The example Voyager bib has four MFHDs but only three MFHDs display.)

https://vufind3.carli.illinois.edu/vf-uiu/Record/UIUdb.3556566

**stos AM**   _from 852 $b and $i_
**v. no.**  _from 853 $a and $b_
**6556253355656620160**  _part of raw MARC record?_
**v. no. (**  _from 853 $a and $b and $i_

![screen shot 2017-08-02 at 8 57 16 am](https://user-images.githubusercontent.com/17598357/28877039-c20e370a-7760-11e7-8ed5-b47821c90ec2.png)

**Comment by cedelis** [comment]
@gibsonjc For whatever reason we had the following config setting for Holdings/summary:

summary = "863abiz:866az"

The default setting was commented-out:

https://github.com/CARLI/vufind/blob/carli-release-3.1.4/config/vufind/VoyagerRestful.ini#L269

I do not recall when or why we un-commented it out (note: it was set up this way when I first added the VoyagerRestful.ini template file into our branch in source control). In vufind3-devel, I've set it to:

summary = "866a"

Is this what we want? https://vufind3-devel.carli.illinois.edu/vf-uiu/Record/UIUdb.3556566

**Comment by gibsonjc** [comment]
This section of the VoyagerRestful.ini does not look like it allows for much detail, which is disappointing because holdings are full of details! Is there any other file or place that influences holdings that you know of?

I believe that we want to keep 

summary = "863abiz:866az"

and, in fact, maybe even expand it to include more subfields that are missing, unless we have some other mechanism for displaying these data other than this single "summary" line. 

Changing it as you have to 

summary = "866a"

"fixes" the display problem I reported only because it eliminates the display of *all* those other subfields full of data, and somewhere in that process is how the improper data in the example seems to be sneaking in.

In short, the system is not displaying the line 
summary = "863abiz:866az" 
as defined and it would be a start to have it do that, but the display would still lack a great deal of information, in my opinion, which is likely a new enhancement.  

(I'm curious if we can define other lines for display, and/or can we use more logic?  What's available right here in no way conforms to MFHD display standards.  It seems to be a raw display of the contents of those field/subfield combinations and completely overlooks the display of corresponding label data in 853 subfields, among other things. I will have a closer look at other VuFind 3 sites and could also try asking on the email list.)

**Comment by cedelis** [comment]
@gibsonjc 

Yes, it appears that VuFind (in general, i.e., across all ILS Drivers, not just VoyagerRestful) simply concatenates all of the summary fields/subfields and lists them line by line.  To be more precise, VuFind treats a group of certain holdings fields (which it calls 'textual fields' in source code) similarly. This group of fields consists of 'holdings_notes', 'summary', 'supplements', 'indexes'. See here: https://github.com/CARLI/vufind/blob/release-3.0/module/VuFind/src/VuFind/ILS/Connection.php#L767


When/if there are more than one values for a particular 'textual field', it simply concatenates the values. See here: 

https://github.com/CARLI/vufind/blob/release-3.0/module/VuFind/src/VuFind/ILS/Logic/Holds.php#L150

Therefore, if we are interested in processing any of these 'textual fields' ('holdings_notes', 'summary', 'supplements', 'indexes') in a more sophisticated manner, we would need to somehow "tell" VuFind to ignore the default processing for these fields (we could override method getHoldingsTextFieldNames() to no longer include the sophisticated field,  or more simply we could blank out the config.ini settings for them instead, e.g., summary = ""), then process them ourselves using our own methods, (e.g., similarly to how we created our own custom variable eresource, we could do the same for summary, by overriding processHoldingRow() and displaying it in our custom holdingsils.phtml template file in RecordTab: https://github.com/CARLI/vufind/blob/carli-release-3.1.4/module/CARLI/src/CARLI/ILS/Driver/VoyagerRestful.php#L692 ).

**Comment by gibsonjc** [comment]
EIU - 12/15/2017 open call - show both the enumeration and chronology of holdings data (major stumbling block to implementation)

**Comment by gibsonjc** [comment]
The problem described by this Issue is Ready for Prod.  I will spin off a new Issue for enhancing the display of enumeration and chronology data.

**Comment by cedelis** [comment]
In prod, commit 69a7f2a652008f033ab12baf568e78d6739ceba4 (I accidentally referenced Issue #147 instead of Issue #247 in the comments).

**Comment by gibsonjc** [comment]
I thought this Issue where the MFHD data wasn't being parsed correctly for display was fixed by something else we did a while back, but I have just had two reports of it still occurring in UIUdb.  

Upon closer examination of the example bib we were using in April 2018, UIUdb.3556566, it may be that we didn't fix it but some editing that UIU did to their MFHD actually changed the data and fixed it accidentally, not us.  Reopening this Issue with new examples!

Examples:

https://i-share.carli.illinois.edu/vf-uiu/Record/UIUdb.989310
<img width="512" alt="screen shot 2018-07-17 at 1 41 42 pm" src="https://user-images.githubusercontent.com/17598357/42838824-6db7ec08-89c7-11e8-83b3-58dbd967b44f.png">

![screen shot 2018-07-17 at 1 42 00 pm](https://user-images.githubusercontent.com/17598357/42838838-7329e9e8-89c7-11e8-9f4e-b7284648a41f.png)

<img width="937" alt="screen shot 2018-07-17 at 1 42 31 pm" src="https://user-images.githubusercontent.com/17598357/42838842-7646f698-89c7-11e8-9aa7-d7b80f2fe4b5.png">

![screen shot 2018-07-17 at 1 42 45 pm](https://user-images.githubusercontent.com/17598357/42838852-7977552e-89c7-11e8-964d-1a02c984ceec.png)

https://i-share.carli.illinois.edu/vf-uiu/Record/UIUdb.561088

<img width="924" alt="screen shot 2018-07-17 at 1 43 13 pm" src="https://user-images.githubusercontent.com/17598357/42838859-7f2b3f30-89c7-11e8-8f8f-f1e1e88853b8.png">

![screen shot 2018-07-17 at 1 43 22 pm](https://user-images.githubusercontent.com/17598357/42838866-82b597e0-89c7-11e8-8c73-0c4ec9225189.png)

**Comment by gibsonjc** [comment]
UIU - ...while looking at several serial records just now, I'm seeing some issues with the serials display.

For the serial, English Historical Review, https://i-share.carli.illinois.edu/vf-uiu/Record/UIUdb.989310 the information showing in the Volume Holdings field, it looks like it is no longer showing Enumeration or Chronology information from the MFHD, it looks like it is pulling garbled information from the 852 field (rather from the 853/863 fields or the 866 field).

It was not doing this a few weeks back. I was actually showing this record to another librarian and the volume information from the MFHD record was showing correctly in the Volume Holdings area of the display. But, it seems, at least for this record, it is not pulling back information from the MFHD record for public display in VuFind.

**Comment by gibsonjc** [comment]
Sometimes this also happens to the display of Item records:
https://i-share.carli.illinois.edu/vf-uiu/Record/UIUdb.2637600
![screen shot 2018-08-14 at 10 50 17 am](https://user-images.githubusercontent.com/17598357/44102672-e4717af0-9faf-11e8-9349-d80ae95b2a6e.png)

The MFHD data itself is garbled (or VuFind is somehow garbling it). The 'Copy' data comes from the 852$t from the MFHD: 

The mfhd ID is 3218973 and the record segment that gets parsed is as follows:

'RECORD_SEGMENT' => '00416cy  a22001815| 4500001000800000004000800008005001700016007000300033008003300036014001300069014001300082852004200095853001400137863001700151863002200168863002200190863002200212^^3218973^^2637600^^20060801154732.0^^ta^^9808030u    0   0002uueng1980803^^1 ^_aARZ-8197^^0 ^_9CMF-3617^^1 ^_bstx^_h015.78^_iR76w^_t1**00416cy  a22001815**| 4500001000800000004000800008005001700016007000300033008003300036014001300069014001300082852004200095853001400137863001700151863002200168863002200190863002200212^^3218973^^2637600^^20060801154732.0^^ta^^9808030u    0   0002uueng1980803^^1 ^_aARZ-8197^^0 ^_9CMF-3617^^1 ^_bstx^_h015.78^_iR76w^_t1^_x(Marked by yr.)^^23^_81^_a(year)^^40^_81.4994^_a1971^^40^_81.5000^_a1959-1962^^40^_81.5002^_a1952-1958^^40^_81.5004^_a1947-1951^^^]^_x(Marked by yr.)^^23^_81^_a(year)^^40^_81.4994^_a1971^^40^_81.5000^_a1959-1962^^40^_81.5002^_a1952-1958^^40^_81.5004^_a1947-1951^^^]',

In the raw Voyager MFHD_DATA table, this UIU MFHD ID has two RECORD_SEGMENTS, with SEQNUM 1 and 2, respectively:

**MFHD_ID      LOCATION_ID         SEQNUM**            RECORD_SEGMENT
**3218973          303      1**          00416cy  a22001815| 4500001000800000004000800008005001700016007000300033008003300036014001300069014001300082852004200095853001400137863001700151863002200168863002200190863002200212 3218973 2637600 20060801154732.0 ta 9808030u    0   0002uueng1980803 1 ¬aARZ-8197 0 ¬9CMF-3617 1 ¬bstx¬h015.78¬iR76w¬t1
**3218973          303      2**          ¬x(Marked by yr.) 23¬81¬a(year) 40¬81.4994¬a1971 40¬81.5000¬a1959-1962 40¬81.5002¬a1952-1958 40¬81.5004¬a1947-1951 

It looks like VuFind is repeating the whole first segment when it is reading the second segment, or something like that, and then confusing itself.

**Comment by cedelis** [comment]
This bug was introduced in commit 5435efceb333ab7e1feaaf47604798fab2e504bd in fix for https://github.com/CARLI/vufind/issues/242

I applied the fix for 247 again.

It's now in TEST.

**Comment by gibsonjc** [comment]
Yes!  All previous examples with problems are corrected on Test now and I don't see anything else like this at the moment. Ready for Prod.

**Comment by cedelis** [comment]
In PROD.
